### PR TITLE
GH-29: Add syntax documentation

### PR DESCRIPTION
**Pull request by Forsinge**
Resolves: GH-29

This adds a reference document for tags and smart punctuation in ModMark, as well as a description of its concepts and functionality. These documents will likely be updated later on to match changes and developments in the language.

### CODE REVIEW

**Comment by CMDJojo** [comment]
Looks great!
One improvement I might suggest is adding a third column to the tags saying if they recursively parse their bodies for more tags or not (currently only Verbatim doesn't do that, so a note might be just fine). And maybe add Math as a tag as well?

Also, your reference made me notice that we don't have a tracking issue for headings as well, that's something we'll need to add. It isn't covered under #36 since it isn't actually tags per se. I will see what I can do tomorrow! But yeah, great work, it is looking good!

**Review by adelhult — APPROVED** [review]
I think this looks good! I guess we can go ahead and merge it right? 

We can always add more information once we have implemented headings and the modules are more complete.

**Review by CMDJojo — COMMENTED** [review]
I think the document is kinda hard to follow, and I think that if we re-structure it a bit it will be much more true to the specification and easier to follow. I would like some changes before merging, as previousley discussed

**Comment by CMDJojo** [comment]
I think you have gotten to a great start with the documentation, and I think it mainly looks good!

I would suggest changing the structure for some of the documentation in `MODMARK.md`. For example, the section about the modules is kind of split up - first, there is a wall of text describing what a module is in detail, followed by some examples further down, and then, at the end of the document, there is a list of what characters are allowed and not. I also think it is important to introduce terminology as we use it, so we don't say that something is restricted to its delimiters while not introducing what a delimiter is until two paragraphs further down.

I think sections like this will be easier to follow if the introduction of modules are in steps, with examples tightly integrated. For example, the structure may change from this:

## Modules
Modules are invoked by square brackets. Information about the module is kept within the brackets while the content is what follows. Exactly how much is consumed depends on the delimiters.
...
Example 3:
`[mymodule] This is a module`
Example 4:
`[mymodule]{&^ This text is included ^&}`
...
As module name, alphanumerics can be used. As delimiters, all special characters can be used.

My suggestion is to change it to something like this:
## Modules
Modules are programs that run on the text you give them, transforming the text at their discretion. To use, or invoke, a module, encase its name in square brackets and place its content following it: `[url] https://example.com`. The module, in this case `url`, is invoked, consuming its content `https://example.com` up until the end of the line or the next whitespace. ... To pass arguments to the module, configuring its behaviour in this specific instance, place them inside the square brackets: `[url color=pink] https://example.com`.

But you have done a great work! I think the content of the documentation is very good, and the `REFERENCE.md` is perfect, I just think that the content may be delivered in a different way to make it easier to read. If you disagree with my comment, feel free to let me know or merge this into main anyways - as Eli said, we can always make changes later!

**Comment by CMDJojo** [comment]
To more clearly show what I mean, I have used your content and re-structured it a bit over at [cmdjojo/GH-29/0](https://github.com/modmark-org/modmark/tree/cmdjojo/GH-29/0), so that it to me looks clearer. Feel free to rebase those changes to this branch if you feel like they are good, or don't if you don't like them!

**Comment by CMDJojo** [comment]
Sounds great! Even though I think some other sections may be restructured in the same way, this may be done later on. I think we should revise the part where \\ allegedly adds a hard break (it currently yields an escaped backslash), but other than that I think this may be merged. The documentation may be improved upon later on, if see fit. Good work, @Forsinge !

**Review by CMDJojo — APPROVED** [review]

**Comment by CMDJojo** [comment]
Also, I do agree with you that the part with smart punctuation may be moved to the syntax specifications. You may move it there before merging, if you feel like it!

**Review by CMDJojo — APPROVED** [review]

**Comment by CMDJojo** [comment]
It would be great to create a tracking issue for the smart punctuation when this is closed! Just so we don't forget it
